### ISSUE TITLE
Issues trying to simulate passive arrays

### ISSUE THREAD

**Issue by Sivler9** [coreylammie/MemTorch#136]
I've been trying to use MemTorch and I've had issues trying to get it to simulate passive arrays, but it always seems to fail to program the networks if I try to use an arbitrary array as weights, although it seems to do fine when the weights are essentially binary (or ternary when using `mt.bh.Scheme.DoubleColumn`) or of size 1x1.

<details><summary>Example</summary>
<p>

```python
import random

import torch
import numpy as np
import memtorch as mt

from torch import nn

DEV = torch.device('cuda:0' if torch.cuda.is_available() else 'cpu')


def set_all_seeds(seed, deterministic=True):
    random.seed(seed)
    np.random.seed(seed)
    torch.manual_seed(seed)
    torch.cuda.manual_seed(seed)
    if deterministic:
        torch.backends.cudnn.benchmark = False
        torch.use_deterministic_algorithms(True)
        torch.backends.cudnn.deterministic = True


def test():
    inp = torch.randn(2, 3)
    matrix = torch.randn(3, 3)

    layer = nn.Linear(3, 3)  # Also tried as class Net() w/ nn.Linear() inside
    layer.weight.data = matrix.T

    patched_model = mt.mn.patch_model(
        model=layer,
        module_parameters_to_patch=[nn.Linear],
        memristor_model=mt.bh.memristor.LinearIonDrift,
        memristor_model_params={},
        #     'time_series_resolution': 1e-6,
        #     'r_off': 5e7,  # mt.bh.StochasticParameter(loc=5e7, scale=2e5, min=1),
        #     'r_on': 1.4e4,  # mt.bh.StochasticParameter(loc=1.4e4, scale=1e5, min=2),
        # },
        scheme=mt.bh.Scheme.SingleColumn,
        transistor=False,
        # tile_shape=(32, 32),
        # ADC_resolution=8,
        # ADC_overflow_rate=0,
        # quant_method='lineal',
        mapping_routine=mt.naive_map,
        programming_routine=mt.naive_program,
        programming_routine_params={'failure_iteration_threshold': 1000},  # , 'simulate_neighbours': False},
        max_input_voltage=.15,
        source_resistance=50.,
        line_resistance=2.,
    )  # Fails to program any weight that does not end up as 1/r_on or 1/r_off

    patched_model.tune_()
    # ... Test matrix product.


if __name__=='__main__':
    set_all_seeds(621, deterministic=False)
    test()
```

</p>
</details>

This does not seem like the expected behavior, at least not with default values or others used through the documentation, I also tried with the VTEAM, but with a bigger time step than default (because otherwise it seems to take more than a day), and it also could not program successfully.

---

I installed it using `python setup.py install` (`CUDA = True`), and it compiled with only some future deprecation warnings, if I remember correctly (I could not find any useful log file).

I'm using CUDA 11.3 in Ubuntu 22.04, and the following PyPI packages:
<details>
  <summary>Output of `pip freeze` </summary>

  ```
asteval==0.9.26
asttokens==2.0.5
backcall==0.2.0
certifi==2021.10.8
charset-normalizer==2.0.12
cycler==0.11.0
decorator==5.1.1
executing==0.8.3
fonttools==4.33.2
future==0.18.2
idna==3.3
ipython==8.2.0
jedi==0.18.1
joblib==1.1.0
kiwisolver==1.4.2
lmfit==1.0.3
matplotlib==3.5.1
matplotlib-inline==0.1.3
memtorch==1.1.6
ninja==1.10.2.3
nose==1.3.7
numpy==1.22.3
packaging==21.3
pandas==1.4.2
parso==0.8.3
pexpect==4.8.0
pickleshare==0.7.5
Pillow==9.1.0
prompt-toolkit==3.0.29
ptyprocess==0.7.0
pure-eval==0.2.2
Pygments==2.12.0
pyparsing==3.0.8
python-dateutil==2.8.2
pytz==2022.1
requests==2.27.1
scikit-learn==1.0.2
scipy==1.8.0
seaborn==0.11.2
six==1.16.0
sklearn==0.0
stack-data==0.2.0
tednet==0.1.3
tensorly==0.7.0
threadpoolctl==3.1.0
torch==1.11.0+cu113
torchvision==0.12.0+cu113
traitlets==5.1.1
typing_extensions==4.2.0
uncertainties==3.1.6
urllib3==1.26.9
wcwidth==0.2.5
wget==3.2
  ```
</details>

---

I've also found a bug on the LinearIonDrift model, while looking into what could be wrong.

In which
https://github.com/coreylammie/MemTorch/blob/305c3a74f7a5ed08ab12ab1974df62f0df8da469/memtorch/bh/memristor/LinearIonDrift.py#L61

and
https://github.com/coreylammie/MemTorch/blob/305c3a74f7a5ed08ab12ab1974df62f0df8da469/memtorch/bh/memristor/LinearIonDrift.py#L99-L102
seem to contradict
https://github.com/coreylammie/MemTorch/blob/305c3a74f7a5ed08ab12ab1974df62f0df8da469/memtorch/bh/memristor/LinearIonDrift.py#L84

But it did not solve the issue.

I can submit a pull request if you prefer, or in future issues, but I do not have any experience contributing to open source code, so I figured I might as well either ask for permission or just give a heads-up.

Also, I can separate the bug from the general issue in the future if you would prefer that.

**Comment by Sivler9** [comment]
For future reference, if any one else encounter this problem, the above code does produce the expected results for bigger resistance values, instead of the range (0, 1) from `torch.randn(...)`. I've also had problems when using `source_resistance`, `line_resistance` very close to zero, to simulate ideal parameters, due to their addition with the crossbar values, which ended up mask by their conductance due to floating point precision.

The second problems does still need to be addressed, I remember going through it with the debugger and the limit jumping from one extreme to the other repeatedly.
